kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -271,48 +272,273 @@ void KvbAppFilter::readBlockRange(BlockId block_id_start,
+    return 0;
+  }
+  auto val = std::get_if<concord::kvbc::categorization::VersionedValue>(&(opt.value()));
+  if (not val) {
     std::stringstream msg;
-    msg << "An event group for trid \"" << client_id_ << "\" cannot be found";
+    msg << "Failed to convert stored tag-specific event group id for key \"" << key << "\" to versioned value";
     throw std::runtime_error(msg.str());
   }
-  auto val = std::get_if<concord::kvbc::categorization::VersionedValue>(&(*opt));
+  return concordUtils::fromBigEndianBuffer<uint64_t>(val->data.data());
+}
+
+uint64_t KvbAppFilter::getValueFromTagTable(const std::string &key) {
+  const auto opt = rostorage_->getLatest(concord::kvbc::categorization::kExecutionEventGroupTagCategory, key);
+  if (not opt) {
+    std::stringstream msg;
+    msg << "Failed to get global event group id for key " << key;
+    LOG_ERROR(logger_, msg.str());
+    throw std::runtime_error(msg.str());
+  }
+  const auto val = std::get_if<concord::kvbc::categorization::ImmutableValue>(&(opt.value()));
   if (not val) {
     std::stringstream msg;
-    msg << "Couldn't convert stored event group id to versioned value";
+    msg << "Failed to convert stored global event group id for key \"" << key << "\" to immutable value";
+    LOG_ERROR(logger_, msg.str());
     throw std::runtime_error(msg.str());
   }
-  auto last_trid_eg_id = concordUtils::fromBigEndianBuffer<uint64_t>(val->data.data());
+  return concordUtils::fromBigEndianBuffer<uint64_t>(val->data.data());
+}
 
-  if (event_group_id_start > event_group_id_end || event_group_id_end > last_trid_eg_id) {
-    throw InvalidEventGroupRange(event_group_id_start, event_group_id_end);
+// We don't store tag-specific public event group ids and need to compute them at runtime.
+
+// This function returns the oldest tag-specific public event group id that the user can request.
+// Due to pruning, it depends on the oldest public event group and the oldest tag-specific event group available.
+uint64_t KvbAppFilter::oldestTagSpecificPublicEventGroupId(const std::string &client_id) {
+  uint64_t public_oldest = getValueFromLatestTable(kPublicEgIdKeyOldest);
+  uint64_t private_oldest = getValueFromLatestTable(client_id + "_oldest");
+  if (!public_oldest && !private_oldest) return 0;
+  return public_oldest + private_oldest - 1;
+}
+
+// This function returns the newest tag-specific public event group id that the user can request.
+// Note that newest tag-specific event group ids will not be updated by pruning
+uint64_t KvbAppFilter::newestTagSpecificPublicEventGroupId(const std::string &client_id) {
+  uint64_t public_newest = getValueFromLatestTable(kPublicEgIdKeyNewest);
+  uint64_t private_newest = getValueFromLatestTable(client_id + "_newest");
+  return public_newest + private_newest;
+}
+
+std::optional<uint64_t> KvbAppFilter::getNextEventGroupId(uint64_t public_end,
+                                                          uint64_t private_end,
```

Comment:
Why do we need `public_end` and `private_end` as parameters?
The client id is a parameter already and so we can construct the key for the private end and the public key is a constant.
We only need those parameters whenever we generate a new batch. Meaning, successive calls on the same batch don't need this information.
The caller will only get one id at a time and hence, whenever he/she is satisfied there is no need to call anymore.

Can we either query storage ourselves whenever we create a generate a new batch or add those parameters to the `state`?

Reply:
Yes, updated.